File: fx.py
#!/usr/bin/env python3
"""
Live currency conversion to INR for PriceSync.

Designer sites quote prices in USD / CAD / INR; the baseline (Studio East Price)
is in INR. To compare and to approve a final price, foreign live prices are
converted to INR using the *current* exchange rate.

Rates are fetched from a free, no-key endpoint and cached (6h). If the network
is unavailable we fall back to sane recent approximations so the app still works.
"""

import time
import requests

_TTL = 6 * 3600
_CACHE = {"at": 0.0, "rates": {}}
# cur -> INR per 1 unit of cur (rough recent values; only used if the API fails)
_FALLBACK = {"INR": 1.0, "USD": 83.0, "CAD": 61.0, "EUR": 90.0, "GBP": 105.0,
             "AUD": 55.0, "AED": 22.6}


def _fetch():
    """Return {cur: INR_per_unit}. Source gives cur-per-1-INR, so we invert."""
    r = requests.get("https://open.er-api.com/v6/latest/INR", timeout=10)
    r.raise_for_status()
    rates = (r.json() or {}).get("rates", {})
    out = {c: (1.0 / v) for c, v in rates.items() if v}
    out["INR"] = 1.0
    if "USD" not in out:                 # sanity: must have at least USD
        raise ValueError("FX payload missing USD")
    return out


def rates():
    now = time.time()
    if now - _CACHE["at"] > _TTL or not _CACHE["rates"]:
        try:
            _CACHE["rates"] = _fetch()
            _CACHE["at"] = now
        except Exception:
            if not _CACHE["rates"]:
                _CACHE["rates"] = dict(_FALLBACK)
    return _CACHE["rates"]


def rate_of(currency):
    c = (currency or "INR").strip().upper()
    return rates().get(c) or _FALLBACK.get(c, 1.0)


def to_inr(amount, currency):
    """Convert an amount in `currency` to INR. None-safe. INR/unknown -> as-is."""
    if amount is None:
        return None
    c = (currency or "INR").strip().upper()
    if c in ("INR", "", "UNKNOWN", "RS", "₹"):
        return round(float(amount), 2)
    return round(float(amount) * rate_of(c), 2)


def snapshot(currencies=("USD", "CAD", "INR")):
    """Current rates for display in the UI."""
    rr = rates()
    return {c: round(rr.get(c, _FALLBACK.get(c, 1.0)), 4) for c in currencies}


if __name__ == "__main__":
    print(snapshot())
    print("100 USD ->", to_inr(100, "USD"), "INR")

File: web/server/src/fx.js
import axios from "axios";

const TTL = 6 * 3600 * 1000;
// Live market rates are genuinely global (not tenant data) — one shared
// cache, refreshed at most every TTL regardless of how many tenants ask.
let cache = { at: 0, rates: {} };
const FALLBACK = { INR: 1, USD: 83, CAD: 61, EUR: 90, GBP: 105, AUD: 55, AED: 22.6 };

// Per-tenant manual FX overrides (fx_override_usd/fx_override_cad in that
// tenant's meta table) — these must never leak across tenants, so they're
// kept in a Map keyed by mboId instead of a single module-level object.
const overrides = new Map();
export function setOverrides(mboId, o = {}) {
  const next = {};
  for (const [k, v] of Object.entries(o)) {
    const n = Number(v);
    if (Number.isFinite(n) && n > 0) next[String(k).toUpperCase()] = n;
  }
  overrides.set(mboId, next);
}
export function getOverrides(mboId) { return { ...(overrides.get(mboId) || {}) }; }

async function fetchRates() {
  const r = await axios.get("https://open.er-api.com/v6/latest/INR", { timeout: 10000 });
  const rates = (r.data && r.data.rates) || {};
  const out = {};
  for (const [c, v] of Object.entries(rates)) if (v) out[c] = 1 / v;
  out.INR = 1;
  if (!out.USD) throw new Error("FX payload missing USD");
  return out;
}

// marketOnly=true skips the tenant's manual Save-Rates override and returns
// the live/fallback market rate only. The override exists for ONE purpose --
// pricing what gets pushed to Shopify (final price = markup over the
// converted baseline) -- and must never leak into mismatch DETECTION.
// Before this flag existed, editing the override rate silently rewrote every
// usd_convert_brand's base_usd on the next pipeline run (via refreshUsdBaselines),
// flipping thousands of rows between matched/mismatch with no real price change
// behind it. Comparison call sites (finalizeOne, refreshUsdBaselines, saveResult's
// own delta recompute) pass true; push/preview call sites (computeFinal, the
// Review page's live preview, promoteLiveToBase/update_base baseline-reset
// actions) omit it and keep using the override, same as before.
export async function rates(mboId, marketOnly = false) {
  if (Date.now() - cache.at > TTL || !Object.keys(cache.rates).length) {
    try { cache = { at: Date.now(), rates: await fetchRates() }; }
    catch { if (!Object.keys(cache.rates).length) cache = { at: Date.now(), rates: { ...FALLBACK } }; }
  }
  return marketOnly ? { ...cache.rates } : { ...cache.rates, ...(overrides.get(mboId) || {}) };
}

export async function rateOf(mboId, cur, marketOnly = false) {
  const c = (cur || "INR").trim().toUpperCase();
  return (await rates(mboId, marketOnly))[c] || FALLBACK[c] || 1;
}

export async function toInr(mboId, amount, cur, marketOnly = false) {
  if (amount == null) return null;
  const c = (cur || "INR").trim().toUpperCase();
  if (["INR", "", "UNKNOWN", "RS", "₹"].includes(c)) return Math.round(amount * 100) / 100;
  return Math.round(amount * (await rateOf(mboId, c, marketOnly)) * 100) / 100;
}

// Pivots any amount+currency through INR to USD — reuses toInr's passthrough
// for INR/unknown-as-INR and rateOf's same cache/fallback/overrides, so this
// stays consistent with every other currency figure the app shows.
export async function toUsd(mboId, amount, cur, marketOnly = false) {
  if (amount == null) return null;
  const inr = await toInr(mboId, amount, cur, marketOnly);
  if (inr == null) return null;
  return Math.round((inr / (await rateOf(mboId, "USD", marketOnly))) * 100) / 100;
}

// Converts at a specific caller-supplied rate instead of any stored one --
// for a brand's own derived rate (brandRate.js: median of 5 sampled products'
// real INR price vs. that brand's own storefront-displayed USD price), so
// the Live/Final $ figures match what the brand's own site shows, not
// whatever a generic INR/USD source happens to quote that day.
export async function toUsdAtRate(mboId, amount, cur, rate) {
  if (amount == null || !rate) return null;
  const inr = await toInr(mboId, amount, cur, true);
  if (inr == null) return null;
  return Math.round((inr / rate) * 100) / 100;
}

// EUR/GBP included because Review's per-row override currency select offers
// them — without a real rate the client previews those amounts at rate 1.
export async function snapshot(mboId, curs = ["USD", "CAD", "EUR", "GBP", "INR"]) {
  const rr = await rates(mboId);
  const out = {};
  for (const c of curs) out[c] = Math.round((rr[c] || FALLBACK[c] || 1) * 10000) / 10000;
  return out;
}
